### PR TITLE
[GEOS-8042] WFS-T fix for InsertResultsFeatureId order

### PR DESCRIPTION
**Pull request by goudine**
For more information https://osgeo-org.atlassian.net/browse/GEOS-8042

### CODE REVIEW

File: src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java [hunk 1]
```diff
@@ -438,4 +439,94 @@ public void testUpdateWithDifferentPrefix() throws Exception {
...
+                + "<gml:LineString>" + "<gml:coordinates decimal=\".\" cs=\",\" ts=\" \">"
+                + "494475.71056415,5433016.8189323 494982.70115662,5435041.95096618"
+                + "</gml:coordinates>" + "</gml:LineString>"
+                + "</cgf:lineStringProperty>" + "<cgf:id>t0002</cgf:id>"
+                + "</cgf:Lines>" + "</wfs:Insert>"
+                + "<wfs:Insert > " + "<cgf:Lines>" + "<cgf:lineStringProperty>"
+                + "<gml:LineString>" + "<gml:coordinates decimal=\".\" cs=\",\" ts=\" \">"
+                + "494475.71056417,5433016.8189329 494982.70115662,5435041.95096618"
+                + "</gml:coordinates>" + "</gml:LineString>"
+                + "</cgf:lineStringProperty>" + "<cgf:id>t0002</cgf:id>"
```
Comment (jodygarnett) [inline src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java]: These all appear to have the same t002 attribute value; making it hard to tell them apart when testing.

File: src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java [hunk 2]
```diff
@@ -438,4 +439,94 @@ public void testUpdateWithDifferentPrefix() throws Exception {
...
    }
 
+    @Test
+    public void testInsert11Features() throws Exception {
+        String features = "<wfs:Transaction service=\"WFS\" version=\"1.0.0\" "
+                + "xmlns:cgf=\"http://www.opengis.net/cite/geometry\" "
+                + "xmlns:ogc=\"http://www.opengis.net/ogc\" "
+                + "xmlns:wfs=\"http://www.opengis.net/wfs\" "
+                + "xmlns:gml=\"http://www.opengis.net/gml\"> "
+                + "<wfs:Insert > " + "<cgf:Lines>" + "<cgf:lineStringProperty>"
```
Comment (jodygarnett) [inline src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java]: Consider sending in your own id values:

```
<cfg:Lines id="t002">
```

And then we can see if they are preserved by the WFS Transaction - that will only work if the datastore on the backend can preserve ids.

File: src/wfs/src/test/java/org/geoserver/wfs/v1_1/TransactionTest.java [hunk 3]
```diff
@@ -794,4 +794,60 @@ public void testEmptyUpdate() throws Exception {
...
+            + "<cgf:pointProperty>" + "<gml:Point>" + "<gml:pos>20 21</gml:pos>"
+            + "</gml:Point>" + "</cgf:pointProperty>" + "<cgf:id>t0001</cgf:id>"
+            + "</cgf:Points>" + "</wfs:Insert>" + "</wfs:Transaction>";
+
+        // get elements from dom and loop through list
+        Document dom = postAsDOM("wfs", feature);
+        NodeList elementsByTagName = dom.getElementsByTagName("ogc:FeatureId");
+        for (int i=0; i<elementsByTagName.getLength(); i++) {
+            String id = elementsByTagName.item(i).getAttributes().item(0).getNodeValue();
+            assertEquals(id, "new"+i);
```
Comment (jodygarnett) [inline src/wfs/src/test/java/org/geoserver/wfs/v1_1/TransactionTest.java]: I am not completely sure if this is the test you are looking for?

File: src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java [hunk 4]
```diff
@@ -438,4 +439,94 @@ public void testUpdateWithDifferentPrefix() throws Exception {
...
+                + "<gml:LineString>" + "<gml:coordinates decimal=\".\" cs=\",\" ts=\" \">"
+                + "494475.71056415,5433016.8189323 494982.70115662,5435041.95096618"
+                + "</gml:coordinates>" + "</gml:LineString>"
+                + "</cgf:lineStringProperty>" + "<cgf:id>t0002</cgf:id>"
+                + "</cgf:Lines>" + "</wfs:Insert>"
+                + "<wfs:Insert > " + "<cgf:Lines>" + "<cgf:lineStringProperty>"
```
Comment (jodygarnett) [inline src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java]: You are testing a sequence of wfs:Inserts.

You may wish to test a single wfs:Insert with a feature collection of 11 features?

File: src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java [hunk 5]
```diff
@@ -438,4 +439,94 @@ public void testUpdateWithDifferentPrefix() throws Exception {
...
+                + "<gml:LineString>" + "<gml:coordinates decimal=\".\" cs=\",\" ts=\" \">"
+                + "494475.71056417,5433016.8189329 494982.70115662,5435041.95096618"
+                + "</gml:coordinates>" + "</gml:LineString>"
+                + "</cgf:lineStringProperty>" + "<cgf:id>t0002</cgf:id>"
+                + "</cgf:Lines>" + "</wfs:Insert>"
+                + "<wfs:Insert > " + "<cgf:Lines>" + "<cgf:lineStringProperty>"
+                + "<gml:LineString>" + "<gml:coordinates decimal=\".\" cs=\",\" ts=\" \">"
+                + "494475.71056415,5433016.8189329 494982.70115665,5435041.95096618"
+                + "</gml:coordinates>" + "</gml:LineString>"
+                + "</cgf:lineStringProperty>" + "<cgf:id>t0002</cgf:id>"
```
Comment (jodygarnett) [inline src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java]: To help generate this content consider making a private utility method:

```
"<wfs:Insert>"+
cfg( "t001", "494475.71056415,5433016.8189329 494982.70115665,5435041.95096618")
cfg( "t002", ... )+
"</wfs:Insert>"
```

```

**Comment by aaime** [comment]
Hey, chimin in quickly, if you want to ensure the right fid order has been returned testing the fids themselves is not going to help, you get the fids, do a GetFeature in order using them, and then check a unique attribute of the feature to know they are actually in the submission order).
E.g., add a "order" attribute in the GML with values "1", "2", "3" and then check that while looping the GetFeatures with the fids you get back the expected order attribute in each one.

File: src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java [hunk 6]
```diff
@@ -438,4 +439,94 @@ public void testUpdateWithDifferentPrefix() throws Exception {
...
+                + "</cgf:Lines>" + "</wfs:Insert>"
+                + "<wfs:Insert > " + "<cgf:Lines>" + "<cgf:lineStringProperty>"
+                + "<gml:LineString>" + "<gml:coordinates decimal=\".\" cs=\",\" ts=\" \">"
+                + "494475.71056417,5433016.8189329 494982.70115667,5435041.95096610"
+                + "</gml:coordinates>" + "</gml:LineString>"
+                + "</cgf:lineStringProperty>" + "<cgf:id>t0002</cgf:id>"
+                + "</cgf:Lines>" + "</wfs:Insert>"
+                + "<wfs:Insert > " + "<cgf:Lines>" + "<cgf:lineStringProperty>"
```
Comment (jodygarnett) [inline src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java]: So this should be:
```
<cfg:Lines id="t007">
```

File: src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java [hunk 7]
```diff
@@ -438,4 +439,94 @@ public void testUpdateWithDifferentPrefix() throws Exception {
...
+                + "<gml:LineString>" + "<gml:coordinates decimal=\".\" cs=\",\" ts=\" \">"
+                + "494475.710564120,5433016.8189323 494982.70115655,5435041.95096670"
+                + "</gml:coordinates>" + "</gml:LineString>"
+                + "</cgf:lineStringProperty>" + "<cgf:id>t0002</cgf:id>"
+                + "</cgf:Lines>" + "</wfs:Insert>"
+                + "</wfs:Transaction>";
```
Comment (jodygarnett) [inline src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java]: 1. Add one insert with a 11+ features
2. Add an unique attribute value that you can check after insert
3. Use a feature id (not an attribute "id")

File: src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java [hunk 8]
```diff
@@ -438,4 +451,94 @@ public void testUpdateWithDifferentPrefix() throws Exception {
...
+
+        Document dom = postAsDOM("wfs", xml);
+        print(dom);
+        dom = getAsDOM("wfs?request=GetFeature&version=1.0.0&service=wfs&typeNames=gs:bar");
+        print(dom);
+
+        NodeList elementsByTagName = dom.getElementsByTagName("gs:bar");
+        for (int i=0; i<elementsByTagName.getLength(); i++) {
+                String id = elementsByTagName.item(i).getAttributes().item(0).getNodeValue();
+                assertEquals("bar."+i, id );
```
Comment (jodygarnett) [inline src/wfs/src/test/java/org/geoserver/wfs/TransactionTest.java]: This looks positive, does not look like your gml:id is being preserved - but your fix is working.

You can try changing your gml:id to "bar.00" ... "bar.012" (if you are saving into a datastore that can preserve fids).

File: src/wfs/src/test/java/org/geoserver/wfs/v1_1/TransactionTest.java [hunk 9]
```diff
@@ -772,7 +773,7 @@ public void testInsertUseExistingId() throws Exception {
        Document dom = postAsDOM("wfs", xml);
        assertEquals("wfs:TransactionResponse", dom.getDocumentElement().getNodeName());
        XMLAssert.assertXpathExists("//ogc:FeatureId[@fid = 'bar.1234']", dom);
-
+        print(dom);
```
Comment (jodygarnett) [inline src/wfs/src/test/java/org/geoserver/wfs/v1_1/TransactionTest.java]: remove print, our logs are long enough already

**Comment by mtCarto** [comment]
This PR has been updated with the feedback from reviews. 
This PR requires [GEOT-5683](https://osgeo-org.atlassian.net/browse/GEOT-5683) [PR:1522](https://github.com/geotools/geotools/pull/1522) and [GEOT-5684](https://osgeo-org.atlassian.net/browse/GEOT-5684) [PR: 1521](https://github.com/geotools/geotools/pull/1521)

**Pull request merged**